internal/codegen/codegen.go

```diff
@@ -0,0 +1,261 @@
+)
+
+// Builder is the heart of stencil, running it is akin to running
+// stencil. Builder handles fetching stencil dependencies and running
+// the actual templating engine and writing the results to disk. Also
+// handled is the extension framework
+type Builder struct {
+	// dir is the path to write templates to
+	dir string
+
+	// manifest is the service manifest that is being used
+	// for this template render
+	manifest *configuration.ServiceManifest
+
+	// extensions is the extensions host that handles all extensions
+	// exposed to templates for this builder
+	extensions *extensions.Host
+
+	// log is the logger used for logging output
+	log logrus.FieldLogger
+
+	// modules is a list of all modules that this builder is utilizing
+	modules []*modules.Module
+}
+
+// NewBuilder returns a new builder
+func NewBuilder(dir string, log logrus.FieldLogger, s *configuration.ServiceManifest) *Builder {
+	_, err := stencil.LoadLockfile("")
+	if !errors.Is(err, os.ErrNotExist) {
+		log.WithError(err).Warn("failed to load lockfile")
+	}
+
+	return &Builder{
+		dir:        dir,
+		manifest:   s,
+		extensions: extensions.NewHost(),
+		log:        log,
+	}
+}
+
+// Run fetches dependencies of the root modules and builds the layered filesystem,
+// after that GenerateFiles is called to actually walk the filesystem and render
+// the templates. This step also does minimal post-processing of the dependencies
+// manifests
+func (b *Builder) Run(ctx context.Context) ([]string, error) {
+	if err := b.processManifest(); err != nil {
+		return nil, errors.Wrap(err, "failed to process service manifest")
+	}
+
+	var err error
+	b.modules, err = modules.GetModulesForService(ctx, b.manifest)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to process modules list")
+	}
+
+	warnings, err := b.GenerateFiles(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return warnings, b.runPostRunCommands(ctx)
+}
+
+// runPostRunCommands runs the postRunCommands set by
+// dependencies
+func (b *Builder) runPostRunCommands(ctx context.Context) error {
+	b.log.Info("Running post run commands")
+	for _, m := range b.modules {
+		mf, err := m.Manifest(ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, cmdStr := range mf.PostRunCommand {
+			b.log.Infof("- %s", cmdStr.Name)
+
+			//nolint:gosec // Why: That's the literal design.
+			cmd := exec.CommandContext(ctx, "/usr/bin/env", "bash", "-c", cmdStr.Command)
+			cmd.Stdin = os.Stdin
+			cmd.Stderr = os.Stderr
+			cmd.Stdout = os.Stdout
+			if err := cmd.Run(); err != nil {
+				return errors.Wrapf(err, "failed to run post run command for module %q", m.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// setDefaultArguments translates a few manifest values
+// into arguments that can be accessed via stencil.Arg
+func (b *Builder) setDefaultArguments() error {
+	if b.manifest.Arguments == nil {
+		b.manifest.Arguments = make(map[string]interface{})
+	}
+	b.manifest.Arguments["name"] = b.manifest.Name
```

Comment:
should these default arguments be underneath a namespace to avoid collision w/other modules?

Reply:
Hmmmm, yeah I think so. I'll put this under "app"